crates/adapters/src/format/avro/test.rs

```diff
@@ -967,6 +967,95 @@ fn test_confluent_avro_output_indexed<K, V>(
     assert_eq!(expected_output, actual_outputs);
 }
 
+#[test]
+fn test_non_unique_keys() {
+    let schema_str = TestStruct::avro_schema().to_string();
+    let config: AvroEncoderConfig = AvroEncoderConfig {
+        schema: Some(schema_str.clone()),
+        key_mode: Some(AvroEncoderKeyMode::None),
+        ..Default::default()
+    };
+    let consumer = MockOutputConsumer::new();
+
+    let k1 = KeyStruct { id: 1 };
+    let v1 = TestStruct {
+        id: 1,
+        b: true,
+        i: None,
+        s: "foo".to_string(),
+    };
+    let v2 = TestStruct {
+        id: 1,
+        b: false,
+        i: None,
+        s: "bar".to_string(),
+    };
+
+    let mut encoder = AvroEncoder::create(
+        "avro_test_endpoint",
+        &Some(KeyStruct::relation_schema()),
+        &TestStruct::relation_schema(),
+        Box::new(consumer),
+        config,
+        None,
+    )
+    .unwrap();
+
+    let zset = OrdIndexedZSet::from_tuples((), vec![Tup2(Tup2(k1.clone(), v1.clone()), 2)]);
+    let zset = Arc::new(<SerBatchImpl<_, KeyStruct, TestStruct>>::new(zset)) as Arc<dyn SerBatch>;
+
+    encoder.consumer().batch_start(0);
+    let err = encoder.encode(zset.as_batch_reader()).unwrap_err();
+    assert!(err.to_string().contains(
+        r#"Error description: Record {"b":true,"i":null,"id":1,"s":"foo"} is inserted 2 times"#
+    ));
+    encoder.consumer().batch_end();
+
+    let zset = OrdIndexedZSet::from_tuples((), vec![Tup2(Tup2(k1.clone(), v1.clone()), -2)]);
+    let zset = Arc::new(<SerBatchImpl<_, KeyStruct, TestStruct>>::new(zset)) as Arc<dyn SerBatch>;
+
+    encoder.consumer().batch_start(0);
+    let err = encoder.encode(zset.as_batch_reader()).unwrap_err();
+    assert!(err.to_string().contains(
+        r#"Error description: Record {"b":true,"i":null,"id":1,"s":"foo"} is deleted 2 times"#
+    ));
+    encoder.consumer().batch_end();
+
+    let zset = OrdIndexedZSet::from_tuples(
+        (),
+        vec![
+            Tup2(Tup2(k1.clone(), v1.clone()), 1),
+            Tup2(Tup2(k1.clone(), v2.clone()), 1),
+        ],
+    );
+    let zset = Arc::new(<SerBatchImpl<_, KeyStruct, TestStruct>>::new(zset)) as Arc<dyn SerBatch>;
+
+    encoder.consumer().batch_start(0);
+    let err = encoder.encode(zset.as_batch_reader()).unwrap_err();
+    println!("{err}");
+    assert!(err
+        .to_string()
+        .contains(r#"Error description: Multiple new values for the same key."#));
```

Comment:
Should this state what the key is?

Reply:
The key is always printed as part of the message, this just checks that the message contains correct error type.